ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -72,3 +160,203 @@ func (r *RayServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		For(&rayv1alpha1.RayService{}).
 		Complete(r)
 }
+
+func (r *RayServiceReconciler) getRayServiceInstance(ctx context.Context, request ctrl.Request) (*rayv1alpha1.RayService, error) {
+	rayServiceInstance := &rayv1alpha1.RayService{}
+	if err := r.Get(ctx, request.NamespacedName, rayServiceInstance); err != nil {
+		if errors.IsNotFound(err) {
+			rayServiceLog.Info("Read request instance not found error!")
+		} else {
+			rayServiceLog.Error(err, "Read request instance error!")
+		}
+		// Error reading the object - requeue the request.
+		return nil, err
+	}
+	return rayServiceInstance, nil
+}
+
+func (r *RayServiceReconciler) updateState(ctx context.Context, rayServiceInstance *rayv1alpha1.RayService, status rayv1alpha1.ServiceStatus, err error) error {
+	rayServiceInstance.Status.ServiceStatus = status
+	if errStatus := r.Status().Update(ctx, rayServiceInstance); errStatus != nil {
+		return fmtErrors.Errorf("combined error: %v %v", err, errStatus)
+	}
+	return err
+}
+
+func (r *RayServiceReconciler) getOrCreateRayClusterInstance(ctx context.Context, rayServiceInstance *rayv1alpha1.RayService) (*rayv1alpha1.RayCluster, error) {
+	// Update ray cluster
+	rayClusterNamespacedName := types.NamespacedName{
+		Namespace: rayServiceInstance.Namespace,
+		Name:      rayServiceInstance.Name + rayClusterSuffix,
+	}
+
+	rayClusterInstance := &rayv1alpha1.RayCluster{}
+	err := r.Get(ctx, rayClusterNamespacedName, rayClusterInstance)
+
+	if err == nil {
+		rayClusterInstance.Spec = rayServiceInstance.Spec.RayClusterSpec
+
+		r.Log.Info("Update ray cluster spec")
+		if err := r.Update(ctx, rayClusterInstance); err != nil {
+			r.Log.Error(err, "Fail to update ray cluster instance!")
+			// Error updating the RayCluster object.
+			return nil, client.IgnoreNotFound(err)
+		}
+	} else if errors.IsNotFound(err) {
+		r.Log.Info("Not found rayCluster, creating rayCluster!")
+		rayClusterInstance, err = r.constructRayClusterForRayService(rayServiceInstance)
+		if err != nil {
+			r.Log.Error(err, "unable to construct rayCluster from spec")
+			// Error construct the RayCluster object - requeue the request.
+			return nil, err
+		}
+		if err := r.Create(ctx, rayClusterInstance); err != nil {
+			r.Log.Error(err, "unable to create rayCluster for rayService", "rayCluster", rayClusterInstance)
+			// Error creating the RayCluster object - requeue the request.
+			return nil, err
+		}
+		r.Log.V(1).Info("created rayCluster for rayService run", "rayCluster", rayClusterInstance)
+	} else {
+		r.Log.Error(err, "Get request rayCluster instance error!")
+		// Error reading the RayCluster object - requeue the request.
+		return nil, client.IgnoreNotFound(err)
```

Comment:
minor: we are pretty sure it's not `NotFound` error, returning err seems better?